core/src/main/scala/org/http4s/blaze/util/Execution.scala

```diff
@@ -106,7 +107,14 @@ object Execution {
         running = false
       } else {
         try r.run()
-        catch { case e: Throwable => reportFailure(e) }
+        catch {
+          case NonFatal(t) => reportFailure(t)
+          case t: Throwable =>
+            // In case someone catches t against all advice of
+            // NonFatal, leave the trampoline in a good state.
+            next()
```

Comment:
I believe that this should not allocate memory or otherwise mask `t`, though I have not studied the implementation of `ArrayDeque`.  We could arguably do without this, but we would have to remove the new test, which otherwise leaves the ThreadLocal in a state that breaks subsequent tests in a flaky, enraging fashion.

Alternatively, we could clear the ThreadLocal in a catch block that targets fatals in `trampoline#execute`, though we'd then potentially lose other queued tasks and deadlock.

Or we could just not test that fatal is fatal.

Reply:
Well, that didn't work.  Let's see what happens [without the fatal test](https://travis-ci.org/http4s/blaze/builds/544050450), but I'm going to bed.

Reply:
I don't think there's anything sensible we can do here.  Any that are in the queue have already been submitted and expect to run, but we need to throw, so we can't run the rest.  Fatal errors are fatal to the trampoline.